### PR TITLE
Move the re-usable pieces of the CPUBackend into a dedicated LLVMIRCodeGen library

### PR DESCRIPTION
**Pull request by opti-mix**
*Description*:
- Move the re-usable pieces of the CPUBackend into a dedicated `LLVMIRCodeGen` library

- Introduce `LLVMBackend`, `LLVMCompiledFunction` and `LLVMIRGen` types that can be used as superclasses for different LLVM-based backends

  Concrete backends can derive from those classes and override specific parts of their logic to implement target-specific code generation.

- Each LLVM-based backend can have its own backend-specific libjit now

  The backend should just override `getLibjitBitcode` to achieve that.

- Re-factor the `CPUBackend` to use this new `LLVMIRCodeGen` library

*Testing*:
ninja test

This PR is an attempt to implement the suggestion form #2447

### CODE REVIEW

**Comment by opti-mix** [comment]
@tlepley-cadence @aturetsk IIRC, your backends may be based on Glow's CPU backend. You may be interested in taking a look at this PR and provide your feedback.

File: include/glow/LLVMIRCodeGen/LLVMIRGen.h [hunk 1]
```diff
@@ -80,14 +79,15 @@ struct DebugInfo {
...
-                                  llvm::DenseMap<Value *, int> &bufferToArgNum,
-                                  llvm::ArrayRef<llvm::Value *> buffers);
+  virtual void
+  emitDataParallelKernelImpl(llvm::IRBuilder<> &builder,
+                             llvm::ArrayRef<const Instruction *> bundle,
+                             llvm::ArrayRef<llvm::Type *> argTypes,
+                             llvm::DenseMap<Value *, int> &bufferToArgNum,
+                             llvm::ArrayRef<llvm::Value *> buffers);
 
 protected:
```
[the remark this answers is not in the document]
Comment (opti-mix) [reply include/glow/LLVMIRCodeGen/LLVMIRGen.h]: Removed.

**Comment by opti-mix** [comment]
@jfix71 Thanks for the review!

Pushed a new version:
- It addressed most of your comments
- It also introduces `CPULLVMIRGen` for handling of the CPUBackend-specific instructions. It also nicely shows how a backend can override the standard logic in `LLVMIRGen` for specific instructions.

**Comment by opti-mix** [comment]
@jfix71 

> I noticed that many doxygen comments do not mention the function parameters/return val. This seemed to be the case prior to this PR too, though. Maybe would be a good opportunity to update them.

Could you be a bit more specific about what functions you have in mind?

File: include/glow/LLVMIRCodeGen/LLVMBackend.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                    llvm::StringRef networkName) const override;
+  /// @}
+
+  /// \returns the size of metrics collected for a single TraceEvent.
+  virtual size_t getTraceEventDataSize() const override {
+    return sizeof(uint64_t);
+  }
+
+protected:
+  /// Method that creates the LLVM IR generator. This gives the possibility to
```
Comment (jfix71) [inline include/glow/LLVMIRCodeGen/LLVMBackend.h]: Make doxygen style

File: include/glow/LLVMIRCodeGen/LLVMBackend.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  /// create a backend that inherits from the CPU backend, while providing
+  /// a specific version of the LLVM IR generator derived from LLVMIRGen.
+  virtual std::unique_ptr<LLVMIRGen>
+  createIRGen(IRFunction *IR, AllocationsInfo &allocationsInfo) const = 0;
+
+  /// Method that creates a CompiledFunction.
```
Comment (jfix71) [inline include/glow/LLVMIRCodeGen/LLVMBackend.h]: Make doxygen style

File: include/glow/LLVMIRCodeGen/LLVMIRGen.h [hunk 4]
```diff
@@ -80,16 +79,16 @@ struct DebugInfo {
 /// This is a class containing a common logic for the generation of the LLVM IR
 /// from an IRFunction. The primary clients of this class are JITs and bundlers.
 class LLVMIRGen {
-private:
+protected:
   /// Implementation of emitDataParallelKernel where we bound the number of
```
Comment (jfix71) [inline include/glow/LLVMIRCodeGen/LLVMIRGen.h]: Make doxygen style

File: include/glow/LLVMIRCodeGen/LLVMIRGen.h [hunk 5]
```diff
@@ -213,12 +223,14 @@ class LLVMIRGen {
   virtual ~LLVMIRGen() {}
   /// Ctor.
   explicit LLVMIRGen(const IRFunction *M, AllocationsInfo &allocationsInfo,
-                     std::string mainEntryName);
+                     std::string mainEntryName, llvm::StringRef libjitBC);
 
   /// Init the TargetMachine using a given target and code model.
```
Comment (jfix71) [inline include/glow/LLVMIRCodeGen/LLVMIRGen.h]: Make doxygen style

File: include/glow/LLVMIRCodeGen/LLVMIRGen.h [hunk 6]
```diff
@@ -213,12 +223,14 @@ class LLVMIRGen {
...
-  void initTargetMachine(llvm::StringRef T, llvm::CodeModel::Model CM);
+  virtual void initTargetMachine(llvm::StringRef T, llvm::CodeModel::Model CM);
 
   /// Emit LLVM-IR for the instruction \p I, using the builder \p builder.
+  /// Derived classes may want to override this function to implement a
+  /// backend-specific LLVM IR generation logic for some intructions.
   virtual void generateLLVMIRForInstr(llvm::IRBuilder<> &builder,
                                       const glow::Instruction *I);
   /// Emit LLVM-IR for the whole IRFunction.
```
Comment (jfix71) [inline include/glow/LLVMIRCodeGen/LLVMIRGen.h]: Make doxygen style, also below `createGlobalVariables()`

File: lib/Backends/CPU/CPUBackend.h [hunk 7]
```diff
@@ -65,18 +45,15 @@ class CPUBackend : public BackendUsingGlowIR {
...
-  static size_t getTraceEventDataSizeStatic() { return sizeof(uint64_t); }
-  size_t getTraceEventDataSize() const override {
-    return CPUBackend::getTraceEventDataSizeStatic();
-  }
-
 protected:
-  /// Method that creates the LLVM IR generator. This gives the possibility to
-  /// create a backend that inherits from the CPU backend, while providing
-  /// a specific version of the LLVM IR generator derived from LLVMIRGen.
+  virtual std::unique_ptr<CompiledFunction> createCompiledFunction(
```
Comment (jfix71) [inline lib/Backends/CPU/CPUBackend.h]: Wrap these methods in a comment saying implementation of LLVMBackend interface.

**Comment by opti-mix** [comment]
@jfix71 Improved the Doxygen comments as you suggested.

File: lib/LLVMIRCodeGen/LLVMIRGen.cpp [hunk 8]
```diff
@@ -1082,40 +1076,6 @@ void LLVMIRGen::generateLLVMIRForDataParallelInstr(
...
-          createCall(builder, F, {loopCount, val, lhsPtr, pointerNull});
-      auto *destAddr = builder.CreateGEP(builder.getFloatTy(), destPtr,
-                                         loopCount, "buffer.element.addr");
-      builder.CreateStore(stackedOpCall, destAddr);
-    }
-
-    break;
-  }
 
 #undef ARITHMETIC_UNARY_OP_CASE
```
Comment (bertmaher) [inline lib/LLVMIRCodeGen/LLVMIRGen.cpp]: Why is this here?  Can we remove it?

**Comment by bertmaher** [comment]
I bet I'm gonna have fun fixing the Buck build when we sync this internally :-p

File: lib/LLVMIRCodeGen/LLVMIRGen.cpp (hunk 8 above)
Comment (opti-mix) [reply lib/LLVMIRCodeGen/LLVMIRGen.cpp]: Done.

File: include/glow/LLVMIRCodeGen/LLVMBackend.h [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  virtual void save(Function *F, llvm::StringRef outputDir,
+                    llvm::StringRef networkName) const override;
+  /// @}
+
+  /// \returns the size of metrics collected for a single TraceEvent.
+  virtual size_t getTraceEventDataSize() const override {
+    return sizeof(uint64_t);
+  }
+
+public:
```
[the remark this answers is not in the document]
Comment (opti-mix) [reply include/glow/LLVMIRCodeGen/LLVMBackend.h]: Done.

**Comment by opti-mix** [comment]
> Do we also need to update docs and CI configs? I was wondering how this builds without changes to CI, but apparently both CPU and new LLVM support flags are True by default. Which contradicts our `docs/Building.md`:

>> By default Glow builds with only the interpreter backend enabled. To enable support for the JIT and/or OpenCL backends, pass additional options to cmake:

Hmm. Good catch! Seems like we contradicted our own `docs/Building.md` for a while ;-) I don't have any strong opinions here. I'm glad to change it to whatever default we pick.

File: include/glow/LLVMIRCodeGen/LLVMCompiledFunction.h [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+  /// Read trace events out of this func and write them into /p ctx
+  virtual void translateTraceEvents(Context *ctx) const override;
+  ///@}
+  //
+
+protected:
+  /// Load constant tensors from \p ctx into \p weightsAddress, as defined by
+  /// the RuntimeBundle (pre-run).
+  virtual void loadPlaceholders(Context *ctx, uint8_t *weightsAddress);
```
Comment (nickgg) [inline include/glow/LLVMIRCodeGen/LLVMCompiledFunction.h]: I this something we'd want to do on all LLVM IR backends? It's really a CPU Function thing I think.
